### PR TITLE
Integrate TreeStand

### PR DESCRIPTION
**Pull request by DerekStride**
# What

Over the discussion at https://github.com/Shopify/tree_stand/issues/30, it's been suggested that the [TreeStand API](https://github.com/Shopify/tree_stand) be integrated directly into [ruby-tree-sitter](https://github.com/Faveod/ruby-tree-sitter). This PR aims to be the initial integration point between the projects.

I've broken the PR up into several commits, the most important of which are:

## [feat!: Integrate TreeStand](https://github.com/Faveod/ruby-tree-sitter/pull/35/commits/2b0c40cb66c842f84dca669500f0316c09cf48fc) 

Pulls over the TreeStand source code & updates it to match the projects style guide.

As part of this commit the project files are updated including a breaking change requiring ruby >= v3.

## [feat(sorbet): Integrate Sorbet Typechecking](https://github.com/Faveod/ruby-tree-sitter/pull/35/commits/7b96e74fe45f039e19fd840353cf3564fa24b16f) 

Fixes typecheck errors by adding type information to the existing tree-sitter bindings & adds a CI check.

## [chore(style): update ruby-tree-sitter to match rubocop styleguide](https://github.com/Faveod/ruby-tree-sitter/pull/35/commits/e52625e9150258df27c80848c8f16a52b745d313)

Updates the existing code to match the style guide & adds a CI check.

# Strategy

## This PR

These are the changes that will be covered by this PR.

- [x] Copy over TreeStand Code
- [x] Typing the C API w/ Sorbet (potentialy w/ [sord](https://github.com/AaronC81/sord))
  - I implemented handcrafted RBI files in https://github.com/Faveod/ruby-tree-sitter/pull/35/commits/bb5866eecfd120d4e07f60d698aaf2d4b73a2a17
- [x] Ensuring TreeStand testsuite is integrated & passing
- [x] Setup rubocop for style guidelines

## Follow-up

These are tasks that still need to be done but are easily separated into other PRs.

- [x] Documenting the C API w/ YARD (https://github.com/Faveod/ruby-tree-sitter/pull/36)
- [ ] Create CONTRIBUTING.md & merge w/ Development.md
- [ ] Update README.md
- [ ] GitHub workflows to publish documentation site
- [ ] Version bump PR

### CODE REVIEW

**Comment by stackmystack** [comment]
Wow @DerekStride that's amazing. What a pleasant surprise :)

I will try to review and help with the list you provided asap.

Reading diagonally, I just noticed something that is not 100% important, and I really don't want to be petty: in the Gemfile, there's a mix of `'` and `"` for strings … I usually prefer single quotes for non-interpolated strings, and double quotes for interpolated or containing escape chars.

I don't mind changing that norm if tooling can help with style. When I work with C++ and Rust, I used clang-format and rustfmt, is there anything you'd recommend/prefer for Ruby so we can get formatting and style out of the way, regardless of what we decide on?

As for versioning, I agree we should talk about it, and FYI I am not happy with  my initial decision of versioning on top of `tree-sitter`'s versions … I thought it would be a good idea. I don't think it is anymore :) Independent semver all the way?

I would also add another point of discussion: Ruby versions. Currently we have support for ruby 2.7, but AFAICT, it's no longer supported, and everyone should move to 3+. Should we drop it? Especially that I had some issues with it segfaulting for reasons I don't really get. For my use case, I really work with latest ruby, so I wouldn't be sorry if 2.7 is dropped.

**Comment by DerekStride** [comment]
> I just noticed something that is not 100% important, and I really don't want to be petty: in the Gemfile, there's a mix of ' and " for strings … I usually prefer single quotes for non-interpolated strings, and double quotes for interpolated or containing escape chars.
> 
> is there anything you'd recommend/prefer for Ruby so we can get formatting and style out of the way?

I'd suggest setting up [rubocop](https://github.com/rubocop/rubocop) to enforce the style. It's the standard tool for ruby.

> As for versioning, I agree we should talk about it, and FYI I am not happy with my initial decision of versioning on top of tree-sitter's versions … I thought it would be a good idea. I don't think it is anymore :) Independent semver all the way?

I'm all for switching to semantic versioning. After landing this PR should we rollout `v1`?

> I would also add another point of discussion: Ruby versions. Currently we have support for ruby 2.7, but AFAICT, it's no longer supported, and everyone should move to 3+. Should we drop it? Especially that I had some issues with it segfaulting for reasons I don't really get. For my use case, I really work with latest ruby, so I wouldn't be sorry if 2.7 is dropped.

TreeStand uses ruby v3 features so I think dropping support for ruby v2.7 makes sense.

**Comment by stackmystack** [comment]
Back from vacations. Happy new year ! :)

I will review all of this today.

As for the CI, it's OK if the ASAN and Valgrind runs fail. They have been like that for a moment and had more urgent matters to deal with.

The disable-lib-sys part is not fixed 100% so if it fails, there's no problem. Just remove the 2.7 please.

I would like to update the documentation (C/README/Contributing) part, but what's the general etiquette on this? Do you mind it if I push on top of your changes in this MR?

**Comment by stackmystack** [comment]
I've also added you as a maintainer.

**Comment by DerekStride** [comment]
> I would like to update the documentation (C/README/Contributing) part, but what's the general etiquette on this? Do you mind it if I push on top of your changes in this MR? Or you'd rather rebase on top of a separate branch ?

I'm happy working on this branch together and then cleaning it up before merging to main.

**Comment by DerekStride** [comment]
Other than a few tweaks I think I have the high-level work done to integrate the projects. I'm going to comb over the changes and separate them into better commits. Some of the other changes that need to be done to the documentation might be better off in separate PRs like #36

File: bin/setup
Comment (stackmystack) [inline bin/setup:22]: Why are you disabling sys lib for the math parser? Wouldn't it be better if it was instructed by the build requirement?

My only concern for the time being is that we're not going to be able to test it with the different options.

Besides, I don't think the flag is stable as of yet. My latest tests with parsers generated from [here](https://github.com/Faveod/tree-sitter-parsers/) failed spectacularly and I haven't had the time to zoom in on it. (I was testing on my local docker so I don't have logs to share).

File: tree_sitter.gemspec
Comment (stackmystack) [inline tree_sitter.gemspec:24]: I guess I'd have to create that repo?

**Review by stackmystack — CHANGES_REQUESTED** [review]
Thank you Derek for the amazing work you've put towards the merger!

LGTM, except the small comments I put here and there.

Feel free to merge whenever you think it's ready. I will rebase on top of the merger and provide the yardoc and address whatever I can from the remaining issues.

… and I need to learn sorbet and integrate it in my dev env :)

File: tree_sitter.gemspec
Comment (DerekStride) [reply tree_sitter.gemspec:24]: That will be the url for a github pages site for this project. We can wait until we have that setup before adding it to the gemspec though. I'll remove it from the diff

Comment (stackmystack) [reply tree_sitter.gemspec:24]: There's no need to remove it. I added the action to publish the docs in this https://github.com/Faveod/ruby-tree-sitter/pull/36.

**Comment by stackmystack** [comment]
Do you think it's ready for merge (even if the remaining comment I put wasn't addressed, it's uncritical)?
Or do you need more time?

File: bin/setup
Comment (DerekStride) [reply bin/setup:22]: We can remove it, it was just me testing it & I forgot to remove it.

**Review by DerekStride — COMMENTED** [review]
Sorry, I've been busy & haven't had time to follow-up on your comments. Your welcome to fix up the PR & ship it if you have the time, otherwise I'll try to follow-up later this week.

**Comment by stackmystack** [comment]
Thank you Derek, I was busy lately. I am going to rebase and merge PR https://github.com/Faveod/ruby-tree-sitter/pull/36.

**Pull request merged**
